### ISSUE TITLE
Documentation on dark subtraction and in-place operation

### ISSUE THREAD

**Issue by evertrol** [astropy/ccdproc#233]
I'm confused about the following in the [docs](http://ccdproc.readthedocs.org/en/latest/ccdproc/index.html):

> Every function returns a *copy* of the data with the operation performed. If,
> for some reason, you wanted to modify the data in-place, do this:

    >>> image_2 = ccdproc.subtract_dark(image_1, dark, dark_exposure=30*u.second, data_exposure=15*u.second, scale=True)

But I see no obvious keyword or anything else (compared to the example code above it) that lets the operation be performed in-place. Testing this also simply shows different ids for the various CCDData instances). 

Browsing through the code, I can't see anything either (the final calculation is, from what I can judge, in `CCDData._ccddata_arithmetic`:

    result_data = operation(self.data, other_value) 

with no hint of an in-place operation).

Am I reading this all wrong and missing something, or is the documentation off?

**Comment by mwcraig** [comment]
@evertrol -- thanks for catching this, the documentation is incorrect.

In-place operation could conceivably be offered as a feature (though it would remain disabled by default)...how important to you would it be to have that option?

**Comment by evertrol** [comment]
I was mostly browsing the documentation to find out how much of use ccdproc could be to me in the near future (compared to e.g. using `astropy.io.fits` and `numpy` directly); this section just struck me as odd.

I may not need in-place operations right now, but it may be of use for me in the near future (say, within half a year). 
While I don't expect to be balancing the memory requirements such that 2 images (science & dark) fit into memory, but a third resulting image doesn't, I do find a 4 to 5 times speed up when subtracting in-place two large (n = 25 000 000; say a 5000 x 5000 pixel CCD) numpy float arrays (measured on a Macbook Air; a better machine with possibly more CPU-optimised operations may have a different ratio).
I could of course simply manipulate the `data` attribute directly, but that takes away a bit from the niceties that ccdproc offers, as well as a clearer view of the processing steps in a script.

Depending on how much you'd like to do yourself (I'm obviously less familiar with the code base), I'd be willing to put in a PR to get this started (since it will have some use for me at some point). 
My attempt would be to add a `inplace=False` kwarg (to `subtract_dark` and related functions), and in `_ccddata_arithmetic` use `operation(self.data, other_value, self.data)` when this is `inplace == True`; `np.subtract` allows for an `out` argument which results in in-place subtraction. I haven't gone through the ccdproc codebase to see if this would work for all operations that use `_ccddata_arithmetic` though!

**Comment by mwcraig** [comment]
@evertrol - it would be great if you are interested in working on it! I'll take a look in the morning at whether there is a more efficient way of approaching it than manually adding the keyword everywhere, but that would certainly work.

**Comment by evertrol** [comment]
I had a go at this, but I hit an obstacle pretty quickly: since `CCDData` inherits from astropy's `NDDataArray` (and therefore automatically deals nicely with uncertainties and units), it also hands out the basic arithmetic to `NDDataArray` when both sides are `CCDData` objects. Which essentially makes this an Astropy issue.

Potentially, one could work around this, but that means implementing quite a few of the `NDDataArray` goodies ourselves.
Alternatively, this could be brought up as an Astropy ticket, but it seems it could be quite a bit more work to have an `inplace` option for an `NDDataArray` operation. In fact, there are quite a few operations going on under the hood in `NDDataArray._arithmetic()`, potentially negating pretty much any speed-up (a very simple at partly implementing this in `NDDataArray` only showed a marginal 20% speed-up for me).

I'm not sure how best to proceed with this.

**Comment by mwcraig** [comment]
@evertrol  -- sounds like a separate ``inplace`` option might not be worth the extra effort for 20%. I'm not too worried about it being in astropy because I'm one of the lead devs for the `NDData` subpackage. 

I think what we originally intended to put in the docs was something like this:

```python
image_1 = ccdproc.subtract_dark(image_1, dark, dark_exposure=30*u.second, data_exposure=15*u.second, scale=True)
```

so that ``image_1`` is overwritten by the result of subtracting the dark. That would be an easy doc fix to make as a first step.

**Comment by evertrol** [comment]
The `image_1 = ...` was a thought that occurred, but since that is a replace, not an in-place operation, I got confused what actually is meant.

**Comment by mwcraig** [comment]
Right, observed too many night in a row and forgot `in_place != replace`.

How would suggest we proceed? I'm ok with a doc wording fix and example fix, or removing that snippet of docs, and/or something more extensive.

**Comment by evertrol** [comment]
Well, I don't really feel it's my call, but I would just drop the sentence/paragraph, to avoid confusion.
And perhaps in some far future, I'll have another look at `NDDataArray._arithmetic()` and see if things can be sped up more using in-place operations, but indeed, I don't really expect it will be worthwhile.

**Comment by mwcraig** [comment]
@evertrol -- OK, will drop. Thanks for the feedback! If you want to open a PR to do the fix, feel free -- I won't get to it for a few days.

**Comment by mwcraig** [comment]
Closed by #237
